### PR TITLE
sparse gradcheck: retire gradcheck_semantics

### PR DESCRIPTION
**Pull request by nikitaved**
Once the tests are parametrized to properly reflect domains/co-domains of sparse functions, success with `masked=False` implies success with `masked=True`, so there is no need in running gradcheck twice.

This PR only affects the tests, it does not touch any gradcheck code.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99298
* __->__ #99043
* #98490
* #104341

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### CODE REVIEW

**Comment by pearu** [comment]
@nikitaved I think you have missed the point of these tests that this PR removes and the previous PR modifies. These tests cover the functionality of `gradcheck` for `masked` being `True` or `False`. When removing `masked=True` checks, you essentially leave the corresponding paths in `gradcheck` untested.

Also, notice that if a test passes with `masked=False`, this does not always imply that the test will pass with `masked=True`. This implication is true only when the sparse inputs are full tensors (`_nnz()` == `numel()` when non-hybrid, for instance) because the densification (that is applied when `masked=False`) will be a no-op. Otherwise, when input tensors have unspecified elements, the gradcheck with `masked=True` has subtleties that are not covered with `masked=False`.

In fact, I think the general implication is: if a test passes with `masked=True` then it will pass with `masked=False`. So, I find this and the previous PR unacceptable from a point of view of reducing the test coverage of gradcheck.
From the operations point of view, indeed, checking with `masked=False` will be sufficient in the domain of non-masked semantics but we'll still need to test the masked semantics as this is what ops in `torch.sparse` namespace implement and a few operations in `torch` namespace support.

**Comment by nikitaved** [comment]
After an offline discussion with @pearu we did reach consensus. Namely, it would be great to have `masked=True/False` to test the logic in gradcheck. It looks, like, @pearu , these cases are covered in `test/test_autograd.py`, so we should be safe here...

**Comment by pearu** [comment]
> It looks, like, @pearu , these cases are covered in `test/test_autograd.py`, so we should be safe here...

Hmm, can you be more specific? AFAIK, test_autograd.py does not cover all the combinations of `to_dense(masked_grad=...)` and `gradcheck(..., masked=..., fast_mode=...)` usages (although ciflow/periodic will cover the `fast_mode` variations).

**Comment by nikitaved** [comment]
@pearu, for example, have a look at https://github.com/pytorch/pytorch/blob/0157b2d722b3411721814cf92467dc32f16f5a56/test/test_autograd.py#L4721.

**Comment by pearu** [comment]
@nikitaved 

> https://github.com/pytorch/pytorch/commit/0157b2d722b3411721814cf92467dc32f16f5a56

this covers only the COO layout but there are no tests with sparse compressed layouts.

**Comment by nikitaved** [comment]
@pearu , none of the tests I modified in `test_sparse.py` concern any compressed formats.

File: test/test_sparse.py [hunk 1]
```diff
@@ -4472,16 +4459,15 @@ def test_to_dense(self, from_layout, device, dtype, index_dtype):
...
+    def test_gradcheck_to_dense(self, from_layout, device, dtype, index_dtype):
         for t in self.generate_simple_inputs(
                 from_layout, device=device, dtype=dtype, index_dtype=index_dtype):
             batch_dim = t.dim() - t.dense_dim() - t.sparse_dim()
             if batch_dim > 0:
                 # TODO: implement batch support in _convert_indices_from_csr_to_coo
                 continue
             t = t.clone().detach().requires_grad_(True)
-            r = gradcheck(lambda x: torch.Tensor.to_dense(x, masked_grad=gradcheck.masked), t)
+            r = gradcheck(lambda x: torch.Tensor.to_dense(x, masked_grad=False), t)
```
Comment (nikitaved) [inline test/test_sparse.py:4470]: Ah, ok, this one touches compressed formats. Will update it, @pearu .
